DigitalLearningSolutions.Data/Services/UserService.cs

```diff
@@ -182,46 +190,50 @@ List<DelegateUser> delegateUsers
         }
 
         // TODO HEEDLS-887 Make sure this logic is correct with the new account structure
-        public void UpdateUserAccountDetailsForAllVerifiedUsers(
+        public void UpdateUserAccountDetailsForAllUsers(
+            int userId,
             MyAccountDetailsData myAccountDetailsData,
             CentreAnswersData? centreAnswersData = null
         )
         {
-            var (verifiedAdminUser, verifiedDelegateUsers) =
-                GetVerifiedLinkedUsersAccounts(
+            var (adminUser, delegateUsers) =
+                GetLinkedUsersAccounts(
                     myAccountDetailsData.AdminId,
-                    myAccountDetailsData.DelegateId,
-                    myAccountDetailsData.Password
+                    myAccountDetailsData.DelegateId
                 );
 
-            if (verifiedAdminUser != null)
+            var detailsLastChecked = DateTime.Now;
+            userDataService.UpdateUserDetailsLastChecked(detailsLastChecked, userId);
+
+            if (adminUser != null)
             {
                 userDataService.UpdateAdminUser(
                     myAccountDetailsData.FirstName,
                     myAccountDetailsData.Surname,
                     myAccountDetailsData.Email,
                     myAccountDetailsData.ProfileImage,
-                    verifiedAdminUser.Id
+                    adminUser.Id
                 );
             }
 
-            if (verifiedDelegateUsers.Count != 0)
+            if (delegateUsers.Count != 0)
             {
+                userDataService.UpdateDelegateAccountCentreSpecificDetailsLastChecked(detailsLastChecked, userId);
                 userDataService.UpdateUser(
                     myAccountDetailsData.FirstName,
                     myAccountDetailsData.Surname,
                     myAccountDetailsData.Email,
                     myAccountDetailsData.ProfileImage,
                     myAccountDetailsData.ProfessionalRegistrationNumber,
                     myAccountDetailsData.HasBeenPromptedForPrn,
-                    centreAnswersData.JobGroupId,
-                    1 // TODO HEEDLS-887 This needs correcting to the correct UserId for the delegate record.
+                    centreAnswersData!.JobGroupId,
+                    userId
                 );
 
                 var oldDelegateDetails =
-                    verifiedDelegateUsers.SingleOrDefault(u => u.Id == myAccountDetailsData.DelegateId);
+                    delegateUsers.SingleOrDefault(u => u.Id == myAccountDetailsData.DelegateId);
 
-                if (oldDelegateDetails != null && centreAnswersData != null)
+                if (oldDelegateDetails != null)

```

Comment:
We're still doing some unnecessary stuff here. Why do we need to fetch the oldDelegateUserDetails from the database when we've already fetched them from the database by user ID earlier? I think we should try to get the delegate user out of the previous list by id, and then only update the delegate account if it is found.

Reply:
@VladNicol In case you miss this comment that I've unresolved

Reply:
I kept that `oldDelegateUserDetails` because `userDataService.GetDelegateAccountsByUserId` returns a `IEnumerable<DelegateAccount>`, while `groupsService.SynchroniseUserChangesWithGroups` takes as argument a `DelegateUser`. Do we have a nice way to map a `DelegateAccount` to a `DelegateUser`? They have almost the same fields, in cases like this I am used to AutoMapper but we don't seem to use that in the project.

Reply:
We don't have a way to map them, but I do wonder whether changing the group synchronization so that it works with the Delgate accounts would be something in scope here. Otherwise I do think it would be good to at some point at least, since I expect we will probably end up with DelegateAccounts in the other place this is called. @SteveJacksonSoft Do we have a ticket to resolve this bit of functionality specifically?

Reply:
I don't think there is a nice and easy way around this, at least until `JobGroupId` is removed from `CentreAnswersData`. We have `GroupsService.SynchroniseUserChangesWithGroups` that calls `LinkedFieldHelper.GetLinkedFieldChanges(CentreAnswersData oldAnswers, CentreAnswersData newAnswers, ...)`.

I think we have to first remove `JobGroupId` from `CentreAnswersData` and then create a `DelegateAccount.GetCentreAnswersData` similar to `DelegateUser.GetCentreAnswersData`

Reply:
The group service method still needs the same data, it's just that that data isn't all in the same place anymore, so some renaming / reshuffling needs to be done in the models. I'll add comments in appropriate places.